validator/validator.bal

```diff
@@ -62,3 +131,172 @@ public isolated function validate(json|anydata data, typedesc<anydata>? targetFH
+            string valueName = "";
+            string valueData = "";
+            regexp:Span? value = invalidArrayElementsData[1];
+            if value !is () {
+                valueName = string `Invalid array element '${value.substring()}'`;
+            }
+            //To get the expected type from the error message
+            regexp:Groups? expectedDataFormat = re `should be of type '([^']+)'`.findGroups(data[i]);
+            if expectedDataFormat is regexp:Groups {
+                regexp:Span? dataType = expectedDataFormat[1];
+                if dataType !is () {
+                    valueData = string `Type of element should be '${dataType.substring()}'`;
+                }
+            }
+            errors.push(string `${valueName}. ${valueData}`);
+        }
+
+        //Add parsing logic here for other parser errors
+
+    }
+
+    //Parsing for fhir multitype scenario
+    //The regex captures a '{' enclosed within \n tags which signifies the start of multitype error.
+    regexp:Groups? multitypeMessage = re `\n\s*\{[\s\S]*`.findGroups(message);
+    if multitypeMessage is regexp:Groups {
+        regexp:Span? value = multitypeMessage[0];
+        if value !is () {
+            string capturedString = value.substring();
+            //Splits message into lines based on \n
+            string[] capturedErrors = splitErrorRegex.split(capturedString);
+
+            string valueName = "";
+            //To get the field name from the error message if error is with the value
+            foreach var i in 0 ... capturedErrors.length() - 1 {
+                regexp:Groups? capturedData = re `^\s*field '([^']+)'`.findGroups(capturedErrors[i]);
+                if capturedData is regexp:Groups {
+                    regexp:Span? fieldData = capturedData[1];
+                    if fieldData !is () {
+                        valueName = fieldData.substring();
+                        errors.push(string `The field '${valueName}' should be of type value[x] or url[x] where x is a valid fhir data type`);
+                        break;
+                    }
+                }
+            }
+            if (valueName === "") {
+                //To get the field name from the error message if error is with the field itself
+                foreach var i in 0 ... capturedErrors.length() - 1 {
+                    regexp:Groups? capturedData = re `^\s* value of field '([^']+)'`.findGroups(capturedErrors[i]);
+                    if capturedData is regexp:Groups {
+                        regexp:Span? fieldData = capturedData[1];
+                        if fieldData !is () {
+                            valueName = fieldData.substring();
+                            errors.push(string `The field '${valueName}' should be of type value[x] or url[x] where x is a valid fhir data type`);
+                            break;
+                        }
+                    }
+                }
+            }
```

Comment:
logic seems same rather than the change in the regex. Shall we move to a util function?